### PR TITLE
feat: configurable and more aggressive cleanup of old AnalysisRuns and Experiments

### LINKED ISSUE #1214: More aggressive cleanup of old AnalysisRuns and Experiments
# Summary

The current default behavior for deleting old AnalysisRuns and Experiments, is that a Rollout will keep around the old objects for the same amount of `spec.revisionHistoryLimit`, which defaults to 10. This seems to be too much for users, who don't really care to keep these around that long. 

I think the default should be changed to delete the old objects when the Rollout reaches Healthy (a.k.a. Completed) state. This will declutter the namespace and things like the Argo UI.

If we change the default to delete old objects more aggressively, one question is if should we provide a knob to increase the retention of old AnalysisRuns/Experiments (e.g. for debugging purposes)? 

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

### PR DESCRIPTION
**Pull request by perenesenko**
Fixes #1214

Signed-off-by: Andrii Perenesenko <andrii.perenesenko@gmail.com>

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: examples/analysis-templates.yaml [hunk 1]
```diff
@@ -36,6 +36,7 @@ metadata:
 spec:
   metrics:
   - name: random-fail
+    count: 2
```
Comment (huikang) [inline examples/analysis-templates.yaml:39]: is this change intentional?

File: pkg/apis/rollouts/v1alpha1/types.go [hunk 2]
```diff
@@ -284,6 +286,14 @@ type CanaryStrategy struct {
 	ScaleDownDelayRevisionLimit *int32 `json:"scaleDownDelayRevisionLimit,omitempty" protobuf:"varint,12,opt,name=scaleDownDelayRevisionLimit"`
 }
 
+// Analysis configuration for the analysis runs to retain
+type Analysis struct {
+	// SucceedRunHistoryLimit limits the number of old analysis runs succeeded to be retained in a history
+	SucceedRunHistoryLimit *int32 `json:"succeedRunHistoryLimit,omitempty" protobuf:"varint,1,opt,name=succeedRunHistoryLimit"`
+	// FailedRunHistoryLimit limits the number of old analysis runs failed to be retained in a history
+	FailedRunHistoryLimit *int32 `json:"failedRunHistoryLimit,omitempty" protobuf:"varint,2,opt,name=failedRunHistoryLimit"`
```
Comment (huikang) [inline pkg/apis/rollouts/v1alpha1/types.go:294]: Could you add the field descriptions to the spec doc: https://github.com/argoproj/argo-rollouts/blob/master/docs/features/specification.md

File: utils/analysis/filter.go [hunk 3]
```diff
@@ -148,6 +163,19 @@ func FilterAnalysisRunsToDelete(ars []*v1alpha1.AnalysisRun, olderRSs []*appsv1.
 			arsToDelete = append(arsToDelete, ar)
 			continue
 		}
+
+		if ar.Status.Phase == v1alpha1.AnalysisPhaseSuccessful {
+			if retainedSucceed < limitSucceedArs {
+				retainedSucceed++
+				continue
+			}
+		} else {
```
Comment (huikang) [inline utils/analysis/filter.go:172]: Shall we distinguish the analysis status between Failed and Running?

File: examples/analysis-templates.yaml (hunk 1 above)
Comment (perenesenko) [reply examples/analysis-templates.yaml:39]: Yes, I've put the count there as there is an error during applying rollout
> AnalysisTemplate random-fail has metric random-fail which runs indefinitely

File: pkg/apis/rollouts/v1alpha1/types.go (hunk 2 above)
Comment (perenesenko) [reply pkg/apis/rollouts/v1alpha1/types.go:294]: Done

File: utils/analysis/filter.go (hunk 3 above)
Comment (perenesenko) [reply utils/analysis/filter.go:172]: Here we're filtering the `otherArs`. As for my understanding, the other (means old) are finished already so they can have succeeded or failed state.

So here I'm checking for 2 subsets of ARs (`ar.Status.Phase == v1alpha1.AnalysisPhaseSuccessful`) and `else` means failed or error.

@huikang What do you think, is that OK or I have to check explicitly for the error/failed in an `else` block?

File: utils/analysis/filter_test.go [hunk 4]
```diff
@@ -196,14 +196,67 @@ func TestFilterAnalysisRunsToDelete(t *testing.T) {
...
+	}
+
+	arS0 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS1 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS2 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS3 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arS4 := ar("a", v1alpha1.AnalysisPhaseSuccessful)
+	arF0 := ar("a", v1alpha1.AnalysisPhaseFailed)
+	arF1 := ar("a", v1alpha1.AnalysisPhaseFailed)
+	arF2 := ar("a", v1alpha1.AnalysisPhaseFailed)
```
Comment (huikang) [inline utils/analysis/filter_test.go:234]: Can we mix the Successful and Failed analysis to make the test more robust? Thanks.

File: utils/defaults/defaults.go [hunk 5]
```diff
@@ -14,7 +14,9 @@ const (
 	// DefaultReplicas default number of replicas for a rollout if the .Spec.Replicas is nil
 	DefaultReplicas = int32(1)
 	// DefaultRevisionHistoryLimit default number of revisions to keep if .Spec.RevisionHistoryLimit is nil
-	DefaultRevisionHistoryLimit = int32(10)
+	DefaultRevisionHistoryLimit           = int32(10)
+	DefaultAnalysisRunSucceedHistoryLimit = int32(1)
+	DefaultAnalysisRunFailedHistoryLimit  = int32(1)
```
Comment (huikang) [inline utils/defaults/defaults.go:19]: Missing comments above the constants?

File: utils/analysis/filter_test.go (hunk 4 above)
Comment (perenesenko) [reply utils/analysis/filter_test.go:234]: Do you mean shuffle the order of ARs?

File: utils/defaults/defaults.go (hunk 5 above)
Comment (perenesenko) [reply utils/defaults/defaults.go]: done

File: utils/analysis/filter.go [hunk 6]
```diff
@@ -108,14 +110,23 @@ func SortAnalysisRunByPodHash(ars []*v1alpha1.AnalysisRun) map[string][]*v1alpha
...
+
 // FilterAnalysisRunsToDelete returns a list of analysis runs that should be deleted in the cases where:
 // 1. The analysis run has no pod hash label,
 // 2. There is no ReplicaSet with the same pod hash as the analysis run
 // 3. The ReplicaSet that has the same pod hash as the analysis run has a deletiontimestamp.
 // Note: It is okay to use pod hash for filtering since the analysis run's pod hash is originally derived from the new RS.
 // Even if there is a library change during the lifetime of the analysis run, the ReplicaSet's pod hash that the analysis
 // run references does not change.
-func FilterAnalysisRunsToDelete(ars []*v1alpha1.AnalysisRun, olderRSs []*appsv1.ReplicaSet) []*v1alpha1.AnalysisRun {
+func FilterAnalysisRunsToDelete(ars []*v1alpha1.AnalysisRun, olderRSs []*appsv1.ReplicaSet, limitSucceedArs int32, limitFailedArs int32) []*v1alpha1.AnalysisRun {
```
[the remark this answers is not in the document]
Comment (perenesenko) [reply utils/analysis/filter.go]: done

File: utils/analysis/filter_test.go (hunk 4 above)
Comment (huikang) [reply utils/analysis/filter_test.go:234]: Yep

Comment (perenesenko) [reply utils/analysis/filter_test.go:234]: done

File: utils/defaults/defaults.go [hunk 7]
```diff
@@ -53,6 +57,22 @@ func GetRevisionHistoryLimitOrDefault(rollout *v1alpha1.Rollout) int32 {
 	return *rollout.Spec.RevisionHistoryLimit
 }
 
+// GetAnalysisRunSucceedHistoryLimitOrDefault returns the specified number of succeed AnalysisRuns to keep or the default number
+func GetAnalysisRunSucceedHistoryLimitOrDefault(rollout *v1alpha1.Rollout) int32 {
+	if rollout.Spec.Analysis == nil || rollout.Spec.Analysis.SucceedRunHistoryLimit == nil {
+		return DefaultAnalysisRunSucceedHistoryLimit
```
Comment (huikang) [inline utils/defaults/defaults.go:63]: Add some unit test to pass the `codecov` test?

File: rollout/analysis.go [hunk 8]
```diff
@@ -3,6 +3,7 @@ package rollout
 import (
 	"context"
 	"fmt"
+	"github.com/argoproj/argo-rollouts/utils/defaults"
```
Comment (huikang) [inline rollout/analysis.go:6]: a space is missing

**Review by huikang — COMMENTED** [review]
Some minor comments; otherwise it looks good.

File: rollout/analysis.go (hunk 8 above)
Comment (perenesenko) [reply rollout/analysis.go:6]: Hi @huikang
Sorry, what space do you mean?

File: utils/defaults/defaults.go (hunk 7 above)
Comment (perenesenko) [reply utils/defaults/defaults.go:63]: done

File: rollout/analysis.go (hunk 8 above)
Comment (huikang) [reply rollout/analysis.go:6]: I meant a new line before line 6 to pass the lint? Sorry about the confusion

Comment (perenesenko) [reply rollout/analysis.go:7]: done

File: pkg/apis/rollouts/v1alpha1/types.go [hunk 9]
```diff
@@ -295,6 +297,14 @@ type CanaryStrategy struct {
 	AbortScaleDownDelaySeconds *int32 `json:"abortScaleDownDelaySeconds,omitempty" protobuf:"varint,13,opt,name=abortScaleDownDelaySeconds"`
 }
 
+// Analysis configuration for the analysis runs to retain
+type Analysis struct {
+	// SucceedRunHistoryLimit limits the number of old analysis runs succeeded to be retained in a history
+	SucceedRunHistoryLimit *int32 `json:"succeedRunHistoryLimit,omitempty" protobuf:"varint,1,opt,name=succeedRunHistoryLimit"`
```
Comment (jessesuen) [inline pkg/apis/rollouts/v1alpha1/types.go:303]: Can we rename this to: `successfulRunHistoryLimit`  in order to be consistent with the Rollout phase `Successful`

```golang
// AnalysisPhase is the overall phase of an AnalysisRun, MetricResult, or Measurement
type AnalysisPhase string

// Possible AnalysisPhase values
const (
	AnalysisPhasePending      AnalysisPhase = "Pending"
	AnalysisPhaseRunning      AnalysisPhase = "Running"
	AnalysisPhaseSuccessful   AnalysisPhase = "Successful"
	AnalysisPhaseFailed       AnalysisPhase = "Failed"
	AnalysisPhaseError        AnalysisPhase = "Error"
	AnalysisPhaseInconclusive AnalysisPhase = "Inconclusive"
)

```

File: docs/features/specification.md [hunk 10]
```diff
@@ -11,6 +11,11 @@ spec:
   # Number of desired pods.
   # Defaults to 1.
   replicas: 5
+  analysis:
+    # limits the number of failed analysis runs to be stored in a history
+    succeedRunHistoryLimit: 0
+    # limits the number of succeeded analysis runs to be stored in a history
+    failedRunHistoryLimit: 0
```
Comment (jessesuen) [inline docs/features/specification.md:18]: 1. We need to document the default behavior when these are omitted
2. We should document that this covers both AnalysisRuns and Experiments
3. It seems you mixed up the descriptions of successful vs. failed
4. Does failedRunHistoryLimit also cover Errors? If so should we name this to `unsuccessfulRunHistoryLimit`?

File: pkg/apis/rollouts/v1alpha1/types.go [hunk 11]
```diff
@@ -295,6 +297,14 @@ type CanaryStrategy struct {
 	AbortScaleDownDelaySeconds *int32 `json:"abortScaleDownDelaySeconds,omitempty" protobuf:"varint,13,opt,name=abortScaleDownDelaySeconds"`
 }
 
+// Analysis configuration for the analysis runs to retain
+type Analysis struct {
```
Comment (jessesuen) [inline pkg/apis/rollouts/v1alpha1/types.go:301]: Nit: The type name of `Analysis` is a little too broad. I think it'll be better rename to `AnalysisStrategy`, since these will hold options on how to do analysis.

NOTE: I'm only talking about the golang type name and not the json field name.

**Review by jessesuen — CHANGES_REQUESTED** [review]

File: pkg/apis/rollouts/v1alpha1/types.go (hunk 9 above)
Comment (perenesenko) [reply pkg/apis/rollouts/v1alpha1/types.go:303]: Renamed *succeed* to *successful* everywhere for this feature

File: docs/features/specification.md (hunk 10 above)
Comment (perenesenko) [reply docs/features/specification.md:18]: 1. Defaults. Done
2. TODO
3. Done
4. I fixed this. Now it covers the failed only.

Comment (perenesenko) [reply docs/features/specification.md]: @jessesuen This feature is covering the AnalysisRuns only.

Maybe we can rename this ticket for AnalysisRuns only, and create another one for the experiments?
For the experiments we need to create other configuration properties I think `analysis.successfulLimit` it's only for analysis

File: pkg/apis/rollouts/v1alpha1/types.go (hunk 11 above)
Comment (perenesenko) [reply pkg/apis/rollouts/v1alpha1/types.go]: Renamed

**Comment by perenesenko** [comment]
On standup was decided to reuse the same parameters `analysis.*` for the Experiments cleaning

**Comment by perenesenko** [comment]
On a standup also discussed about the changing `analysis.failed...` => `analysis.unsuccessful...` and that parameter should includes all 3 phases: **Failed, Error, Inconclusive**

File: utils/analysis/filter.go [hunk 12]
```diff
@@ -4,6 +4,8 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 
 	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+
+	"sort"
```
[the remark this answers is not in the document]
Comment (perenesenko) [reply utils/analysis/filter.go:8]: Hi @huikang, is there any specific rules for the import sorting?
I just put it in the abc order

File: pkg/apis/rollouts/v1alpha1/types.go [hunk 13]
```diff
@@ -295,6 +297,15 @@ type CanaryStrategy struct {
 	AbortScaleDownDelaySeconds *int32 `json:"abortScaleDownDelaySeconds,omitempty" protobuf:"varint,13,opt,name=abortScaleDownDelaySeconds"`
 }
 
+// AnalysisStrategy configuration for the analysis runs to retain
```
[the remark this answers is not in the document]
Comment (perenesenko) [reply pkg/apis/rollouts/v1alpha1/types.go]: Done

File: docs/features/specification.md (hunk 10 above)
Comment (perenesenko) [reply docs/features/specification.md]: 1. Done
2. Documented these params covers both AnalysisRuns and Experiments. Done
3. Done
4. Implement `unsuccessfulRunHistoryLimit`. Done

File: utils/analysis/filter.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (huikang) [reply utils/analysis/filter.go:8]: I don't think there is any in this project. But when I autosave the go file, the IDE typically puts those built-int package on the top and then sort by the package name. I think this is a minor issue we can ignore.

**Comment by huikang** [comment]
LGTM

File: utils/analysis/filter.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jessesuen) [reply utils/analysis/filter.go:8]: > Hi @huikang, is there any specific rules for the import sorting?

Golang imports blocks are conventionally:
1. standard libraries
2. external libraries
3. libraries from same project package (i.e. github.com/argoproj/argo-rollouts/*)

**Review by jessesuen — APPROVED** [review]
